R/crunch-data-frame.R

```diff
@@ -0,0 +1,292 @@
+        }
+        row_inds <- i
+    }
+    return(row_inds)
+}
+
+#' @export
+#' @rdname CrunchDataFrame
+`[[.CrunchDataFrame` <- function (x, i) {
+    col_name <- col_name_from_index(x, i)
+    
+    return(get_CDF_var(col_name, x, mode = attr(x, "mode")))
+}
+
+#' @export
+#' @rdname CrunchDataFrame
+`[[<-.CrunchDataFrame` <- function (x, i, value) {
+    col_name <- col_name_from_index(x, i)
+    
+    x <- set_CDF_var(col_name = col_name, cdf = x, value = value)
+    return(x)
+}
+
+col_name_from_index <- function(x, i) {
+    if (is.character(i)) {
+        col_name <- i
+        # check if any don't exist
+    } else if (is.numeric(i)) {
+        col_name <- names(x)[i]
+    } else {
+        halt("column subsetting must be done with either numeric or ",
+             "a character")
+    }
+    return(col_name)
+}
+
+#' @export
+#' @rdname CrunchDataFrame
+`$.CrunchDataFrame` <- function (x, i) {
+    return(get_CDF_var(i, x, mode = attr(x, "mode")))
+}
+
+#' @export
+#' @rdname CrunchDataFrame
+`$<-.CrunchDataFrame` <- function (x, i, value) {
+    x <- set_CDF_var(col_name = i, cdf = x, value = value)
+    return(x)
+}
+
+get_CDF_var <- function (col_name, cdf, ...) {
+    if (!is(cdf, "CrunchDataFrame")) { 
+        halt("The cdf argument must be a CrunchDataFrame, got ", class(cdf), " instead.")
+    }
+    if (!col_name %in% names(cdf)) {
+        halt("The variable ", dQuote(col_name), " is not found in the CrunchDataFrame.")
+    }
+    
+    return(get(col_name, cdf)) # need [ord] here?
+}
+
+get_var_from_server <- function (col_name, cdf, ...) {
+    ord <- attr(cdf, "order")
+    ds <- attr(cdf, "crunchDataset")
+    if (!is.null(ord)) {
+        return(
+            as.vector(ds[[col_name]][ord], ...)[match(ord, sort(unique(ord)))]
+        )
+    } else {
+        # no reordering if order is null
+        return(as.vector(ds[[col_name]], ...))
+    }
+}
+
+set_CDF_var <- function (col_name, row_inds, cdf, value) {
+    if (!is(cdf, "CrunchDataFrame")) { 
+        halt("The cdf argument must be a CrunchDataFrame, got ", class(cdf), " instead.")
+    }
+    
+    # remove the variable if the value is NULL
+    if (is.null(value) & missing(row_inds)) {
+        return(invisible(rm_CDF_var(col_name = col_name, cdf = cdf)))
+    }
+    
+    if (missing(row_inds)) {
+        row_inds <- seq_len(nrow(cdf))
+    }
+    
+    # validate row lengths
+    if (length(value) == 1 & length(row_inds) > 1) {
+        # if there is a single value, but more than one row, then we need to 
+        # repeat the value for as many times as there are rows. This is the 
+        # only exception to the rule that there must be the same number of 
+        # values as rows.
+        value <- rep(value, length(row_inds))
+    } else if (is.data.frame(value)) {
+        # we got a data.frame as values, likelythis is an array or MR varaible
+        # unclass does what data.frame(v1=data.frame(a=1, b=2)) would do
```

Comment:
This comment doesn't make sense to me/doesn't seem to match the code. And/or you have left in commented code that should be deleted?

Reply:
I've harmonized this behavior with our discussion earlier. Array-like variables are now flattened.